### ISSUE TITLE
Bugfix: Retry on "database locked" error ( or switch to SQL)

### ISSUE THREAD

**Issue by bash-stack** [ArchiveBox/ArchiveBox#601]
when using the documented method to add or update links in parallel , 
or using the gui while updating on the cli ,
plenty of `database locked` errors , 
and since the somehow lengthy ( min. 500ms per page only to check , up to 6*120 s with timeout 2m )
a simple `RETRY_DB_WRITE` variable and the corresponding code would help 

(or just switch to SQL , since it is way faster , especially when running 20Mbyte+ sqlite files from SD/MMC )

**Comment by pirate** [comment]
By "switch to SQL" do you mean switch to PostgreSQL or MySQL? If so we're probably not going to do that anytime soon. You can already do it optionally but I'm not going to make it the default since it massively increases the complexity for the average user.

Parallel archiving is not yet fully stable, it's going to get better in v0.5 and v0.6, you can see the roadmap here: https://github.com/ArchiveBox/ArchiveBox/wiki/Roadmap

The plan is to enable parallel archiving by moving to a message passing architecture and using an async queue worker like Huey/Dramatiq to handle running multiple extractor threads / link archiving threads at once.

**Comment by bash-stack** [comment]
you should have a look at [cabot](https://cabotapp.com/use/configuration.html) , it is in python  , has a que 
and does nearly the same ( get something from somewhere and analyze  headers/content/statuscode )

**Comment by bash-stack** [comment]
> You can already do it optionally

How does one do that ?

**Comment by pirate** [comment]
You can edit `archivebox/core/settings.py` and change the `DATABASE` section to use a standard postgres connection.

Although I highly doubt it will be faster, SQLite is extremely fast on SSDs, and often beats PostgreSQL on benchmarks for queries run locally.

**Comment by asitemade4u** [comment]
Just to throw in my two cents, it is actually not only a matter of performance.

Postgres is IMO a better choice for a whole lot of reasons:
- better handling of concurrent requests
- overall, better maintainability and security
- and for me at least: ability to seamlessly connect ArchiveBox database to our other databases

Could you direct me to a relevant documentation for that "standard django postgres connection" you mentioned?
Thanks,
S.

**Comment by pirate** [comment]
While Postgres is better at concurrent *writes* (not necessarily reads), it's definitely not a clearcut winner on "better at maintainability and security". Postgres is significantly less tested than SQLite. I think you forget that Airbus planes, rockets, boats, most iOS apps and other native apps, and tons of modern critical infrastructure run on SQLite!

- https://www.sqlite.org/testing.html
- https://blog.expensify.com/2018/01/08/scaling-sqlite-to-4m-qps-on-a-single-server/
- https://antonz.org/sqlite-is-not-a-toy-database/
- https://github.com/rqlite/rqlite

We might add Postgres/MySQL support in the future, but it's not a priority because ArchiveBox is focused on long-term data format durability above all else, and SQLite is the clear winner in that regard, having been standardized to the point of wide adoption far beyond Postgres and MySQL and others. Side note, I love Postgres, but honestly MySQL is slightly better for the ArchiveBox use-case, InnoDB compression is better, and it can handle many more concurrent threads than Postgres with lower resource usage on smaller machines/containers.

**Comment by pirate** [comment]
Note I've added a new DB/filesystem troubleshooting area to the wiki that may help people arriving here from Google: https://github.com/ArchiveBox/ArchiveBox/wiki/Upgrading-or-Merging-Archives#database-troubleshooting

Contributions/suggestions welcome there.
